fix(blurhash): Use preview API to generate the previews

artonge opened the pull request:
**1. feat(previews): Support in memory preview request**
This allows callers to use the API without increasing the disk usage.

**2. fix(blurhash): Use preview API to generate the previews**
This allows to benefit from all the checks done by the API.
This also used the newly introduced `cacheResult` argument to limit disk usage.

- Fix github.com/nextcloud/server/issues/52063

artonge committed:
feat(previews): Support in memory preview request

This allows callers to use the API without increasing the disk usage.

Example: blurhash generation, where we request a preview for all uploaded pictures, but don't want to necessarily store that preview.
Signed-off-by: Louis Chemineau <louis@chmn.me>

skjnldsv commented on lib/private/Blurhash/Listener/GenerateBlurhashMetadata.php:
```diff
@@ -64,7 +70,9 @@ public function handle(Event $event): void {
 		}
 
-		$image = $this->resizedImageFromFile($file);
+		$preview = $this->preview->getPreview($file, 256, 256, cacheResult: false);
```
wasn't it checked lately that 64 was enough to barely see the difference with blurhashes? :)

artonge replied:
Oh, yes, yes, yes, let's not revert that, thanks!

skjnldsv reviewed (approved):
One comment otherwise good!

skjnldsv reviewed (approved):
One comment otherwise good!

come-nc commented:
psalm:update-baseline needed

miaulalala commented:
/backport to stable31

miaulalala commented:
/backport to stable30

miaulalala commented:
/backport to stable30

miaulalala commented:
/backport to stable29

max-nextcloud commented on lib/private/Blurhash/Listener/GenerateBlurhashMetadata.php:
```diff
@@ -64,7 +70,9 @@ public function handle(Event $event): void {
 		}
 
-		$image = $this->resizedImageFromFile($file);
+		$preview = $this->preview->getPreview($file, 64, 64, cacheResult: false);
```
The size was 30x30 before. That's what the `RESIZE_BOXSIZE` constant is used for.

```suggestion
		$preview = $this->preview->getPreview($file, self::RESIZE_BOXSIZE, self::RESIZE_BOXSIZE, cacheResult: false);
```

I think 64 would also be okay - but maybe we can drop the constant then / change it.

max-nextcloud replied:
https://github.com/nextcloud/server/pull/50494 has screenshots with 30px vs 300px.

susnux commented:
:test_tube: :boom: 

> 1) Test\Preview\GeneratorTest::testGetNewPreview
> Expectation failed for method name is "putContent" when invoked 1 time(s).
> Method was expected to be called 1 times, actually called 0 times.

artonge committed:
fix(blurhash): Use preview API to generate the previews

This allows to benefit from all the checks done by the preview API.
This also use the newly introduced `cacheResult` argument to limit disk usage.

Signed-off-by: Louis Chemineau <louis@chmn.me>

skazi0 commented:
> > @artonge did you test this with `'enable_previews' => false` ?
> 
> No. By reading the code, I assume this would skip blurhash generation. Which would be expected as blurhashes are a kind of preview.

It doesn't generate blurhash but also logs:
```
"Exception":"OCP\\\\Files\\\\NotFoundException","Message":"Previews disabled","Code":0,
"Trace":[
{"file":"/var/www/nextcloud/lib/private/PreviewManager.php","line":157,"function":"throwIfPreviewsDisabled","class":"OC\\\\PreviewManager","type":"->"},
{"file":"/var/www/nextcloud/lib/private/Blurhash/Listener/GenerateBlurhashMetadata.php","line":71,"function":"getPreview","class":"OC\\\\PreviewManager","type":"->"},
{"file":"/var/www/nextcloud/lib/private/EventDispatcher/ServiceEventListener.php","line":68,"function":"handle","class":"OC\\\\Blurhash\\\\Listener\\\\GenerateBlurhashMetadata","type":"->"},
{"file":"/var/www/nextcloud/3rdparty/symfony/event-dispatcher/EventDispatcher.php","line":220,"function":"__invoke","class":"OC\\\\EventDispatcher\\\\ServiceEventListener","type":"->"},
{"file":"/var/www/nextcloud/3rdparty/symfony/event-dispatcher/EventDispatcher.php","line":56,"function":"callListeners","class":"Symfony\\\\Component\\\\EventDispatcher\\\\EventDispatcher","type":"->"},
{"file":"/var/www/nextcloud/lib/private/EventDispatcher/EventDispatcher.php","line":67,"function":"dispatch","class":"Symfony\\\\Component\\\\EventDispatcher\\\\EventDispatcher","type":"->"},
{"file":"/var/www/nextcloud/lib/private/EventDispatcher/EventDispatcher.php","line":79,"function":"dispatch","class":"OC\\\\EventDispatcher\\\\EventDispatcher","type":"->"},
{"file":"/var/www/nextcloud/lib/private/FilesMetadata/FilesMetadataManager.php","line":100,"function":"dispatchTyped","class":"OC\\\\EventDispatcher\\\\EventDispatcher","type":"->"},
{"file":"/var/www/nextcloud/lib/private/FilesMetadata/Job/UpdateSingleMetadata.php","line":43,"function":"refreshMetadata","class":"OC\\\\FilesMetadata\\\\FilesMetadataManager","type":"->"},
{"file":"/var/www/nextcloud/lib/public/BackgroundJob/Job.php","line":61,"function":"run","class":"OC\\\\FilesMetadata\\\\Job\\\\UpdateSingleMetadata","type":"->"},
{"file":"/var/www/nextcloud/lib/public/BackgroundJob/QueuedJob.php","line":43,"function":"start","class":"OCP\\\\BackgroundJob\\\\Job","type":"->"},
{"file":"/var/www/nextcloud/lib/public/BackgroundJob/QueuedJob.php","line":29,"function":"start","class":"OCP\\\\BackgroundJob\\\\QueuedJob","type":"->"},
{"file":"/var/www/nextcloud/cron.php","line":170,"function":"execute","class":"OCP\\\\BackgroundJob\\\\QueuedJob","type":"->"}
]
```
